apps/vscode/src/providers/editor/toggle.ts

```diff
@@ -12,16 +12,88 @@
  * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
  *
  */
-
+import * as vscode from "vscode";
 import { commands, window, workspace, TextDocument, ViewColumn } from "vscode";
+import * as path from 'path';
+import * as quarto from "quarto-core";
+import fs from "node:fs";
+import yaml from "js-yaml";
 import { Command } from "../../core/command";
 import { isQuartoDoc, kQuartoLanguageId } from "../../core/doc";
 import { VisualEditorProvider } from "./editor";
 
 
+export async function determineMode(doc: string): Promise<string | undefined> {
+  let editorOpener = undefined;
+
+  // check if file itself has a mode
+  if (hasEditorMode(doc, "source")) {
+    editorOpener = "source";
+  }
+  else if (hasEditorMode(doc, "visual")) {
+    editorOpener = "visual";
+  }
+  // check if has a _quarto.yml or _quarto.yaml file with editor specified
+  else {
+    editorOpener = workspaceHasQuartoYaml();
+  }
+
+  return editorOpener;
+}
 
+export async function setEditorOpener() {
+  const config = vscode.workspace.getConfiguration('quarto').get<string>('defaultEditor');
+  const viewType = config === 'visual' ? VisualEditorProvider.viewType : 'textEditor';
+  vscode.workspace.getConfiguration('workbench').update('editor.defaultView', viewType, true);
+  await vscode.commands.executeCommand("workbench.action.setDefaultEditor",
+    vscode.Uri.file('filename.qmd'),
```

Comment:
Actually I wonder if we could expose `updateUserAssociations()` with the slightly better name of `setEditorAssociations()`, but then expose it from `positron.d.ts` rather than as a command (either in the top level `'positron'` namespace, or under a new `'workbench'` sub namespace, mirroring what VS Code has. That would work like our existing `'language'` and `'runtime'` sub namespaces).

That would probably be nice because it would force you to look at Quarto's `hasHooks()` and `hooksApi()` helpers to determine whether or not you are in positron (we use that in a few other places too)

(FWIW `saveAllTitled` had to be a command to be able to call it from R, but this feels more like an Extension API feature that can be a "real" function exported from `positron.d.ts`)